### PR TITLE
applications: nrf5340_audio: Implement broadcast receiver

### PR DESCRIPTION
**Pull request by erikrobstad**
In the new architecture.
Implemented using separate
streamctrl file.

### CODE REVIEW

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * @return	0 for success, error otherwise.
+ */
+int broadcast_sink_stop(void);
+
+/**
+ * @brief	Enable the LE Audio broadcast sink.
+ *
+ * @return	0 for success, error otherwise.
+ */
+int broadcast_sink_enable(le_audio_receive_cb recv_cb);
```
Comment (alexsven) [inline applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h]: No documentation for recv_cb

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 2]
```diff
@@ -584,17 +448,80 @@ static int zbus_subscribers_create(void)
...
+	return 0;
+}
+
+/**
+ * @brief	Zbus listener to receive events from bt_mgmt.
+ *
+ * @note	Will in most cases be called from BT_RX context,
+ *		so there should not be too much processing done here.
+ */
+static void bt_mgmt_evt_handler(const struct zbus_channel *chan)
```
Comment (alexsven) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:450]: Missing doc for chan

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 3]
```diff
@@ -628,16 +566,23 @@ int streamctrl_start(void)
...
 
-	ret = bt_content_ctrl_init();
-	ERR_CHK(ret);
+	ret = audio_datapath_thread_create();
+	ERR_CHK_MSG(ret, "Failed to create audio datapath thread");
 
-	ARG_UNUSED(le_audio_rx_data_handler);
+	ret = broadcast_sink_enable(le_audio_rx_data_handler);
+	if (ret) {
+		LOG_WRN("Failed to enable broadcast sink");
```
Comment (alexsven) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Could be ERR I guess, same for the one below

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 4]
```diff
@@ -584,17 +448,80 @@ static int zbus_subscribers_create(void)
...
+ */
+static void bt_mgmt_evt_handler(const struct zbus_channel *chan)
+{
+	int ret;
+	const struct bt_mgmt_msg *msg;
+
+	msg = zbus_chan_const_msg(chan);
+	uint8_t event = msg->event;
+
+	if (event == BT_MGMT_PA_SYNC_OBJECT_READY) {
```
Comment (alexsven) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:458]:
```suggestion
	if (msg->event == BT_MGMT_PA_SYNC_OBJECT_READY) {
```

Comment (alexsven) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:458]: This is probably true for all the other cases of zbus usage in this file and in streamctrl_broadcast_source

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 5]
```diff
@@ -584,17 +448,80 @@ static int zbus_subscribers_create(void)
 		return ret;
 	};
```
Comment (alexsven) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]:
```suggestion
	}
```

Comment (alexsven) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: I see the same in streamctrl_broadcast_source

**Review by alexsven — DISMISSED** [review]

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 6]
```diff
@@ -608,10 +535,21 @@ static int zbus_subscribers_create(void)
...
 
+uint8_t stream_state_get(void)
+{
+	return strm_state;
+}
+
+void streamctrl_send(void const *const data, size_t size, uint8_t num_ch)
+{
+	ARG_UNUSED(data);
+	ARG_UNUSED(size);
```
Comment (rick1082) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:536]: Should we add a WRN/ERR in here for pointing out this is not used?

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 7]
```diff
@@ -628,16 +566,23 @@ int streamctrl_start(void)
 	ret = zbus_subscribers_create();
 	ERR_CHK_MSG(ret, "Failed to create zbus subscriber threads");
 
-	ret = bt_rend_init();
-	ERR_CHK(ret);
+	ret = zbus_link();
```
Comment (rick1082) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Nitpick: Would we use zbus_setup() or zbus_init() instead of link?

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 8]
```diff
@@ -197,68 +191,9 @@ static void stream_state_set(enum stream_state stream_state_new)
...
-			} else {
-				LOG_WRN("Problem with sending LE audio data, ret: %d", ret);
-			}
-		}
-
-		prev_ret = ret;
-	}
-}
-
-#define TEST_TONE_BASE_FREQ_HZ 1000
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:227]: I may see this later. Where is the test tone now?

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 9]
```diff
@@ -322,26 +252,11 @@ static void button_msg_sub_thread(void)
...
-					break;
-				}
-
-				ret = test_tone_start();
-				if (ret) {
-					LOG_WRN("Failed to play test tone, ret: %d", ret);
-				}
-
-				break;
+			ret = broadcast_sink_change_active_audio_stream();
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:243]: Can we name this more explicitly? It is not clear if the channel or broadcaster is changed.

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 10]
```diff
@@ -445,34 +362,23 @@ static void le_audio_msg_sub_thread(void)
...
-			 * }
-			 *
-			 * LOG_DBG("Sampling rate: %d Hz", sampling_rate_hz);
-			 *LOG_DBG("Bitrate: %d bps", bitrate_bps);
-			 */
-			break;
+			ret = broadcast_sink_config_get(&bitrate_bps, &sampling_rate_hz,
+							&pres_delay_us);
+			if (ret) {
+				LOG_WRN("Failed to get config: %d", ret);
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:356]: WRN or ERR?

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 11]
```diff
@@ -584,17 +448,80 @@ static int zbus_subscribers_create(void)
...
+	uint8_t event = msg->event;
+
+	if (event == BT_MGMT_PA_SYNC_OBJECT_READY) {
+		LOG_INF("PA sync object ready");
+		ret = broadcast_sink_pa_sync_set(msg->pa_sync, msg->broadcast_id);
+		if (ret) {
+			LOG_WRN("Failed to set PA sync");
+		}
+	} else {
+		LOG_WRN("Unexpected/unhandled bt_mgmt event: %d", event);
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:465]: Smells like LOG_ERR

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 12]
```diff
@@ -584,17 +448,80 @@ static int zbus_subscribers_create(void)
...
+}
+
+ZBUS_LISTENER_DEFINE(bt_mgmt_evt_listen, bt_mgmt_evt_handler);
+
+/**
+ * @brief	Link zbus producers and observers.
+ *
+ * @return	0 for success, error otherwise.
+ */
+static int zbus_link(void)
```
Comment (koffes) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]:
```suggestion
static int zbus_add_observers(void)
```

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c [hunk 13]
```diff
@@ -297,7 +297,8 @@ static void base_recv_cb(struct bt_bap_broadcast_sink *sink, const struct bt_bap
 			((active_stream_index == AUDIO_CH_L) ? CH_L_TAG : CH_R_TAG));
 		LOG_DBG("Waiting for syncable");
 	} else {
-		LOG_WRN("Found no suitable stream");
+		LOG_DBG("Found no suitable stream");
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c:300]: Why was this changed to a DBG?

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c [hunk 14]
```diff
@@ -430,7 +431,7 @@ static int bis_headset_cleanup(void)
 	return 0;
 }
 
-static int change_active_audio_stream(void)
+int broadcast_sink_change_active_audio_stream(void)
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c:434]: Ref. comment above. "stream" is a bit indescriptive.

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+/**
+ * @brief	Disable the LE Audio broadcast sink.
+ *
+ * @return	0 for success, error otherwise.
+ */
+int broadcast_sink_disable(void);
```
Comment (koffes) [inline applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h]: Nitpicky: Often, we have the init/start/enable functions at the bottom of the file. This goes for the source file as well.

**Review by koffes — CHANGES_REQUESTED** [review]

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 16]
```diff
@@ -493,11 +399,12 @@ static void le_audio_msg_sub_thread(void)
...
 
 		case LE_AUDIO_EVT_NO_VALID_CFG:
-			LOG_WRN("No valid configurations found, will disconnect");
+			LOG_WRN("No valid configurations found, disabling the broadcast sink");
 
-			ret = bt_mgmt_conn_disconnect(msg.conn, BT_HCI_ERR_REMOTE_USER_TERM_CONN);
+			ret = broadcast_sink_disable();
 			if (ret) {
-				LOG_ERR("Failed to disconnect: %d", ret);
+				LOG_ERR("Failed to disable the broadcast sink");
```
Comment (rick1082) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Let's print the ret in here.

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 17]
```diff
@@ -352,11 +267,11 @@ static void button_msg_sub_thread(void)
...
 
 				break;
-			} else if (IS_ENABLED(CONFIG_BT_BAP_BROADCAST_SINK)) {
-				/* Will eventually be handled by different applications */
-				ret = le_audio_disable();
+			} else {
+				ret = broadcast_sink_disable();
 				if (ret) {
-					LOG_ERR("Failed to disable the broadcast receiver");
+					LOG_ERR("Failed to disable the broadcast sink");
```
Comment (rick1082) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Would be great if we print ret in here.

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c [hunk 18]
```diff
@@ -197,68 +191,9 @@ static void stream_state_set(enum stream_state stream_state_new)
...
-
-		ret = audio_encode_test_tone_set(test_tone_hz);
-		ERR_CHK_MSG(ret, "Failed to generate test tone");
-	}
-
-	return 0;
-}
-
-/* Handle button activity */
+/**
```
Comment (erikrobstad) [inline applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:182]: Remove line 114-119

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h (hunk 1 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h]: Added doc

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 2 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:450]: Added doc in both streamctrl_broadcast_sink.c and streamctrl_broadcast_source.c

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 3 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Changed to ERR_CHK_MSG

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 4 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:458]: The variable 'event' is already 'msg->event' (see line 466)

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 5 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Changed in both files

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 6 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:536]: Added a LOG_WRN()

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 7 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Changed name to "zbus_link_producers_observers"

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 8 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:227]: It is not needed for broadcast sink, therefore it is removed

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 10 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:356]: Maybe WRN since audio might still be working even if this function fails. The presentation delay might be wrong though. But I am open for discussion.

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 9 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:243]: Do you have any name suggestions? I cant come up with any better names that are not too long.
Added more explanation to the Doxygen doc.

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 11 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:465]: Hmm I think this could be WRN since the system might work fine even if an unexpected/unhandled event happens. But I am open for discussion.
If we change this to LOG_ERR it must be done in every relevant zbus thread.

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 12 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Ref the earlier comment from Rick I have changed this name to zbus_link_producers_observers()

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c (hunk 13 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c:300]: Since the LE_AUDIO_EVT_NO_VALID_CFG event is now published on BIS headset as well (not only CIS gateway) we get a WRN print in streamctrl

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c (hunk 14 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.c:434]: Same comment as the one above:
Do you have any name suggestions? I cant come up with any better names that are not too long.
Added more explanation to the Doxygen doc.

File: applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h (hunk 15 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/bluetooth/bt_stream/broadcast/broadcast_sink.h]: Agreed, changed :)

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 16 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Done

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 17 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Done

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 18 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c:182]: Done

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 17 above)
Comment (rick1082) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Probably we missed one in here :P

**Review by koffes — APPROVED** [review]

File: applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c (hunk 17 above)
Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Ah, sorry! I thought this comment was a duplicate. I will fix

Comment (erikrobstad) [reply applications/nrf5340_audio/src/audio/streamctrl_broadcast_sink.c]: Done

**Review by rick1082 — APPROVED** [review]
